File: python/895-maximum-frequency-stack.py



from collections import deque


class FreqStack:

    def __init__(self):
        self.frequency_map = {}
        self.hash_map = defaultdict(int)
        self.max_frequency = 0
        

    def push(self, val: int) -> None:
        freq = self.hash_map[val] + 1
        self.hash_map[val] = freq
        
        if freq in self.frequency_map:
            self.frequency_map[freq].append(val)
        else:
            self.frequency_map[freq] = deque([val])
            self.max_frequency = max( self.max_frequency , freq)
        
        
    def pop(self) -> int:
        q = self.frequency_map[self.max_frequency]
        val = q.popleft()
        self.hash_map[val] = self.hash_map[val] - 1
        if not q:
           del self.frequency_map[self.max_frequency]
           self.max_frequency -= 1
        return val
       


# Your FreqStack object will be instantiated and called as such:
# obj = FreqStack()
# obj.push(val)
# param_2 = obj.pop()

File: tao/java/895-maximum-frequency-stack.java
import java.util.HashMap;
import java.util.PriorityQueue;

class FreqStack {
    PriorityQueue<Node> pq;
    HashMap<Integer, Integer> map;
    int globalTime = 0;
    public FreqStack() {
        pq = new PriorityQueue<>();
        map = new HashMap<>();
    }
    
    public void push(int val) {
        int frequency = map.getOrDefault(val,0) + 1;
        Node node = new Node(val, globalTime++, frequency);
        map.put(val, frequency);
        pq.add(node);
    }
    
    public int pop() {
        int num = pq.poll().num;
        map.put(num, map.get(num)-1);
        return num;
    }
}

class Node implements Comparable<Node> {
    int num;
    int time;
    int frequency;
    public Node(int num, int time, int frequency) {
        this.num = num;
        this.time = time;
        this.frequency = frequency;
    }

    @Override
    public int compareTo(Node that) {
        if (that.frequency > this.frequency || (that.frequency == this.frequency && that.time > this.time)) {
            return 1;
        }
        return -1;
    }
}

/**
 * Your FreqStack object will be instantiated and called as such:
 * FreqStack obj = new FreqStack();
 * obj.push(val);
 * int param_2 = obj.pop();
 */